File: mobile/ios/BGMIServerGlitch/BGMIServerGlitch/ViewModels/ServerAccessViewModel.swift
import Foundation
import Combine

class ServerAccessViewModel: ObservableObject {
    @Published var activeAccess: [ServerAccess] = []
    @Published var remainingTime: Int = 0
    @Published var isLoading = false
    @Published var errorMessage: String?
    @Published var timerRunning = false
    
    private var cancellables = Set<AnyCancellable>()
    private let apiClient = APIClient.shared
    private var timer: Timer?
    
    func startServer(serverName: String, season: String, accessType: String = "FREE_TRIAL") {
        isLoading = true
        errorMessage = nil
        
        apiClient.startServer(serverName: serverName, season: season, accessType: accessType)
            .receive(on: DispatchQueue.main)
            .sink { [weak self] completion in
                self?.isLoading = false
                if case .failure(let error) = completion {
                    self?.errorMessage = error.localizedDescription
                }
            } receiveValue: { [weak self] access in
                self?.remainingTime = access.remainingTimeMillis / 1000
                self?.startTimer()
                self?.fetchActiveAccess()
            }
            .store(in: &cancellables)
    }
    
    func fetchActiveAccess() {
        apiClient.getActiveAccess()
            .receive(on: DispatchQueue.main)
            .sink { [weak self] completion in
                if case .failure(let error) = completion {
                    self?.errorMessage = error.localizedDescription
                }
            } receiveValue: { [weak self] accesses in
                self?.activeAccess = accesses
            }
            .store(in: &cancellables)
    }
    
    func stopServer(accessId: Int) {
        apiClient.stopServer(accessId: accessId)
            .receive(on: DispatchQueue.main)
            .sink { [weak self] completion in
                if case .failure(let error) = completion {
                    self?.errorMessage = error.localizedDescription
                }
            } receiveValue: { [weak self] _ in
                self?.fetchActiveAccess()
            }
            .store(in: &cancellables)
    }
    
    private func startTimer() {
        timerRunning = true
        timer = Timer.scheduledTimer(withTimeInterval: 1.0, repeats: true) { [weak self] _ in
            if self?.remainingTime ?? 0 > 0 {
                self?.remainingTime -= 1
            } else {
                self?.stopTimer()
            }
        }
    }
    
    private func stopTimer() {
        timer?.invalidate()
        timer = nil
        timerRunning = false
        remainingTime = 0
    }
    
    deinit {
        stopTimer()
    }
}

File: mobile/android/app/src/main/java/com/bgmi/serverglitch/viewmodel/ServerAccessViewModel.kt
package com.bgmi.serverglitch.viewmodel

import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.bgmi.serverglitch.models.request.StartServerRequest
import com.bgmi.serverglitch.models.response.ServerAccessResponse
import com.bgmi.serverglitch.network.ApiClient
import com.bgmi.serverglitch.network.ApiService
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.delay
import kotlinx.coroutines.isActive
import kotlinx.coroutines.launch
import timber.log.Timber
import javax.inject.Inject

@HiltViewModel
class ServerAccessViewModel @Inject constructor() : ViewModel() {

    private val apiService = ApiClient.create(ApiService::class.java)

    private val _serverAccessResult = MutableLiveData<Result<ServerAccessResponse>>()
    val serverAccessResult: LiveData<Result<ServerAccessResponse>> = _serverAccessResult

    private val _activeAccess = MutableLiveData<List<ServerAccessResponse>>()
    val activeAccess: LiveData<List<ServerAccessResponse>> = _activeAccess

    private val _remainingTime = MutableLiveData<Long>()
    val remainingTime: LiveData<Long> = _remainingTime

    private val _isLoading = MutableLiveData(false)
    val isLoading: LiveData<Boolean> = _isLoading

    fun startServer(serverName: String, season: String, accessType: String = "FREE_TRIAL") {
        _isLoading.value = true
        viewModelScope.launch {
            try {
                val response = apiService.startServer(
                    StartServerRequest(serverName, season, accessType)
                )
                _serverAccessResult.value = Result.success(response)
                _remainingTime.value = response.remainingTimeMillis
                startTimer(response.remainingTimeMillis, response.id)
                Timber.d("Server access started: ${response.serverName}")
            } catch (e: Exception) {
                _serverAccessResult.value = Result.failure(e)
                Timber.e("Server start error: ${e.message}")
            } finally {
                _isLoading.value = false
            }
        }
    }

    fun getActiveAccess() {
        _isLoading.value = true
        viewModelScope.launch {
            try {
                val response = apiService.getActiveAccess()
                _activeAccess.value = response
                Timber.d("Fetched ${response.size} active accesses")
            } catch (e: Exception) {
                Timber.e("Error fetching active access: ${e.message}")
            } finally {
                _isLoading.value = false
            }
        }
    }

    fun stopServer(accessId: Long) {
        viewModelScope.launch {
            try {
                apiService.stopServer(accessId)
                Timber.d("Server access stopped: $accessId")
                getActiveAccess()
            } catch (e: Exception) {
                Timber.e("Error stopping server: ${e.message}")
            }
        }
    }

    private fun startTimer(initialTimeMillis: Long, accessId: Long) {
        viewModelScope.launch {
            var remaining = initialTimeMillis
            while (remaining > 0 && isActive) {
                delay(1000)
                remaining -= 1000
                _remainingTime.value = remaining
                Timber.d("Time remaining: ${remaining / 1000} seconds")
            }
            if (remaining <= 0) {
                stopServer(accessId)
                _remainingTime.value = 0
            }
        }
    }
}
